File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int bestDayToBuy = prices[0];
        int bestDayToSell = prices[0];
        
        int maxProfix = 0;
        
        for (int i=1; i<prices.length; i++){
            if (prices[i] < bestDayToBuy){
                bestDayToBuy = prices[i];
                bestDayToSell = prices[i];
            }
            else if (prices[i] >  bestDayToBuy){
                bestDayToSell = prices[i];
                int tempProfix = bestDayToSell - bestDayToBuy;
                maxProfix = Math.max(maxProfix, tempProfix);
            }
        }
        
        return maxProfix;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        best_day_to_buy:int = prices[0]
        best_day_to_sell:int = prices[0]
        
        profit:int = 0
        
        for price in prices:
            if price < best_day_to_buy:
                best_day_to_buy = price
                best_day_to_sell = price
            elif price > best_day_to_sell:
                best_day_to_sell = price
                
                profit = max((best_day_to_sell - best_day_to_buy), profit)
                # if best_day_to_sell > best_day_to_buy:
                #     res =  best_day_to_sell - best_day_to_buy 
                #     if profit < res:
                #         profit = res
        
        return profit
                